torch/_lobpcg.py

```diff
@@ -0,0 +1,702 @@
+            P = mm(S_, mm(Z[:, n - nc:], _utils.basis(_utils.transpose(Z[:n - nc, n - nc:]))))
+            np = P.shape[-1]
+
+            # check convergence
+            self.update_residual()
+            nc = self.update_converged_count()
+
+            # update S
+            self.S[:, :n] = self.X
+            self.S[:, n:n + np] = P
+            W = self._get_ortho(self.R[:, nc:], self.S[:, :n + np])
+            ns = self.ivars['converged_end'] = n + np + W.shape[-1]
+            self.S[:, n + np:ns] = W
+
+    def _get_rayleigh_ritz_transform(self, S):
+        """Return a transformation matrix that is used in Rayleigh-Ritz
+        procedure for reducing a general eigenvalue problem :math:`(S^TAS)
+        C = (S^TBS) C E` to a standard eigenvalue problem :math: `(Ri^T
+        S^TAS Ri) Z = Z E` where `C = Ri Z`.
+
+        .. note:: In the original Rayleight-Ritz procedure in
+          [DuerschEtal2018], the problem is formulated as follows::
+
+            SAS = S^T A S
+            SBS = S^T B S
+            D = (<diagonal matrix of SBS>) ** -1/2
+            R^T R = Cholesky(D SBS D)
+            Ri = D R^-1
+            solve Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          To reduce the number of matrix products (denoted by empty
+          space between matrices), here we incorporate the diagonal
+          matrix D into Ri and introduce element-wise products
+          (denoted by symbol `*`) so that the Rayleight-Ritz procedure
+          becomes::
+
+            SAS = S^T A S
+            SBS = S^T B S
+            d = (<diagonal of SBS>) ** -1/2    # this is 1-d column vector
+            dd = d d^T                         # this is 2-d matrix
+            R^T R = Cholesky(dd * SBS)
+            Ri = R'^-1 * d                     # broadcasting
+            solve Ri^T SAS Ri Z = Theta Z
+            C = Ri Z
+
+          where `dd` is 2-d matrix that replaces matrix products `D M
+          D` with one element-wise product `M * dd`; and `d` replaces
+          matrix product `D M` with element-wise product `M *
+          d`. Also, creating the diagonal matrix `D` is avoided.
+
+        Arguments:
+        S (Tensor): the matrix basis for the search subspace, size is
+                    :math:`(m, n)`.
+
+        Returns:
+        Ri (tensor): upper-triangular transformation matrix of size
+                     :math:`(n, n)`.
+
+        """
+        B = self.B
+        mm = torch.matmul
+        SBS = _utils.qform(B, S)
+
+        d1 = SBS.diagonal(0, -2, -1) ** -0.5
+        d = d1.reshape(d1.shape[0], 1)
+
+        dd = mm(d, _utils.transpose(d))
+        R = torch.cholesky(dd * SBS, upper=True)
+        # TODO: could use LAPACK ?trtri as R is upper-triangular
+        Rinv = torch.inverse(R)
+        return Rinv * d
+
+    def _get_svqb(self,
+                  U,     # Tensor
+                  drop,  # bool
+                  tau    # float
+                  ):
+        """Return B-orthonormal U.
```

Comment:
Not sure whether it is important but you mean that `U^T B U = I` here?

Reply:
yes.